presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractMapBlock.java

```diff
@@ -335,10 +338,55 @@ public boolean isNull(int position)
         return mapIsNull != null && mapIsNull[position + getOffsetBase()];
     }
 
+    // Only for testing
+    public boolean isHashTablesCreated()
+    {
+        return getHashTables().getRawHashTables() != null;
+    }
+
     private void checkReadablePosition(int position)
     {
         if (position < 0 || position >= getPositionCount()) {
             throw new IllegalArgumentException("position is not valid");
         }
     }
+
+    public static class HashTables
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapBlock.class).instanceSize();
+
+        // Hash to location in map. Writes to the field is protected by "this" monitor.
+        @Nullable
+        private volatile int[] hashTables;
+
+        HashTables()
+        {
+            this.hashTables = null;
+        }
+
+        HashTables(int[] hashTables)
+        {
+            this.hashTables = hashTables;
+        }
+
+        int[] getRawHashTables()
```

Comment:
Why don't we return `Optional<int[]>` here?

```
Optional<int[]> getRawHashTables()
{
    return Optional.ofNullable(hashTables);
}
```

Reply:
I thought we need to do it for public functions only. @wenleix do you also think we need to do this?

Reply:
I don't think using `Optional`  depends on whether a method is public or not, it's a good practice either way.